assets/src/components/site-scan-context-provider/index.js

```diff
@@ -0,0 +1,347 @@
+} ) {
+	const { originalOptions: { theme_support: themeSupport } } = useContext( Options );
+	const { setAsyncError } = useAsyncError();
+	const [ state, dispatch ] = useReducer( siteScanReducer, initialState );
+	const {
+		cache,
+		currentlyScannedUrlIndex,
+		pluginIssues,
+		scannableUrls,
+		stale,
+		status,
+		themeIssues,
+	} = state;
+
+	/**
+	 * Preflight check.
+	 */
+	useEffect( () => {
+		if ( status ) {
+			return;
+		}
+
+		if ( ! validateNonce ) {
+			throw new Error( 'Invalid site scan configuration' );
+		}
+
+		dispatch( { type: ACTION_SCANNABLE_URLS_REQUEST } );
+	}, [ status, validateNonce ] );
+
+	/**
+	 * This component sets state inside async functions. Use this ref to prevent
+	 * state updates after unmount.
+	 */
+	const hasUnmounted = useRef( false );
+	useEffect( () => () => {
+		hasUnmounted.current = true;
+	}, [] );
+
+	const startSiteScan = useCallback( ( args = {} ) => {
+		dispatch( {
+			type: ACTION_SCAN_INITIALIZE,
+			cache: args?.cache,
+		} );
+	}, [] );
+
+	const cancelSiteScan = useCallback( () => {
+		dispatch( { type: ACTION_SCAN_CANCEL } );
+	}, [] );
+
+	/**
+	 * Cancel scan and invalidate current results whenever theme mode changes.
+	 */
+	const previousThemeSupport = usePrevious( themeSupport );
+	useEffect( () => {
+		if ( previousThemeSupport && previousThemeSupport !== themeSupport ) {
+			dispatch( { type: ACTION_SCAN_CANCEL } );
+			dispatch( { type: ACTION_SCAN_INVALIDATE } );
+		}
+	}, [ previousThemeSupport, themeSupport ] );
+
+	/**
+	 * Fetch scannable URLs from the REST endpoint.
+	 */
+	useEffect( () => {
+		( async () => {
+			if ( status !== STATUS_REQUEST_SCANNABLE_URLS ) {
+				return;
+			}
+
+			dispatch( { type: ACTION_SCANNABLE_URLS_FETCH } );
+
+			try {
+				const fields = [ 'url', 'amp_url', 'type', 'label' ];
+				const response = await apiFetch( {
+					path: addQueryArgs( scannableUrlsRestPath, {
+						_fields: fetchCachedValidationErrors ? [ ...fields, 'validation_errors', 'stale' ] : fields,
+					} ),
+				} );
+
+				if ( true === hasUnmounted.current ) {
+					return;
+				}
+
+				dispatch( {
+					type: ACTION_SCANNABLE_URLS_RECEIVE,
+					scannableUrls: response,
+				} );
+			} catch ( e ) {
+				setAsyncError( e );
+			}
+		} )();
+	}, [ fetchCachedValidationErrors, scannableUrlsRestPath, setAsyncError, status ] );
+
+	/**
+	 * Scan site URLs sequentially.
+	 */
+	useEffect( () => {
+		( async () => {
+			if ( status !== STATUS_IDLE ) {
+				return;
+			}
+
+			dispatch( { type: ACTION_SCAN_VALIDATE_URL } );
+
+			try {
+				const urlType = ampFirst || themeSupport === STANDARD ? 'url' : 'amp_url';
+				const url = scannableUrls[ currentlyScannedUrlIndex ][ urlType ];
+				const args = {
+					'amp-first': ampFirst || undefined,
+					amp_validate: {
+						cache: cache || undefined,
+						nonce: validateNonce,
+						omit_stylesheets: true,
+					},
+				};
+				const validationResults = await apiFetch( {
+					url: addQueryArgs( url, args ),
+				} );
```

Comment:
> Unfortunately, I'm still getting stale validation results whenever I'm refreshing the Settings page. I'm not sure why that happens.

It appears this is due to `_locale=user` being added to each of the URLs being requested. This is something that `apiFetch` is doing via the `userLocaleMiddleware`.

Since the request being made here is not technically a REST API request but rather a frontend request, I think replacing this with a simple `fetch` may suffice:

```suggestion
				const validationResults = await fetch( addQueryArgs( url, args ) );
```